# App Store和Google Paly Markets中App的探索

在這項專案中，假設我們在App公司中擔任數據分析師，我們要讓產品上架在Google Play和App Store平台。

公司的產品免費下載安裝，而主要的營收來源為App內的廣告，這表示我們的主要營收會受使用者人數影響－使用者越多，廣告商投放的廣告就越多，進而使營收受益。這項專案的目的是分析資料使研發部門了解哪一種類的App比較能吸引用戶。


如上所述，我們的目標是幫助研發部門了解哪個類別的app比較能吸引到Google Play和App Store的用戶。為了達到這項目的，我們需要搜集與分析兩大平台的資料。

在2018年九月，大約有200萬個app在App Store上架，與210萬個app在Android的Google Play平台上架。

![photo](https://s3.amazonaws.com/dq-content/350/py1m8_statista.png)
<center>來源: [Statista](https://www.statista.com/statistics/276623/number-of-apps-available-in-leading-app-stores/)</center>

搜集超過400萬比資料需要大量時間與金錢，為了避免花費過大的資源，所以我們從大筆資料中取樣，因此我們使用以下兩筆資料集來完成我們的目標。

- 包含大約1萬筆Google Play的Android apps資料集，在2018年八月蒐集，[資料來源](https://dq-content.s3.amazonaws.com/350/googleplaystore.csv)

- 包含大約7千筆App Store的iOS apps資料集，在2017年七月蒐集，[資料來源](https://dq-content.s3.amazonaws.com/350/AppleStore.csv)

### 讀入與探索資料

In [1]:
# 讀入這兩筆資料

from csv import reader

# 檢查App Store資料集
# 去除表頭

opened_file = open('AppleStore.csv')
read_file = reader(opened_file)
ios = list(read_file)
ios_header = ios[0]
ios = ios[1:]


# 檢查Google play資料集
# 去除表頭
opened_file = open('googleplaystore.csv')
read_file = reader(opened_file)
android = list(read_file)
android_header = android[0]
android= android[1:]

In [2]:
# 建立一個計算資料數量的函式
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]
    for row in dataset_slice:
        print(row)
        print('\n') 
    
    if rows_and_columns:
        print('row數: ', len(dataset))
        print('column數: ', len(dataset[0]))

In [3]:
print(android_header)
print('\n')
explore_data(android, 0, 5, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Eve

In [4]:
print(ios_header)
print('\n') #增加可讀性
explore_data(ios, 0, 5, True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']


row數:  7197
column數:  16


In [5]:
print(android_header)
print(android[10472])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


In [6]:
print(len(android))
del android[10472]
print(len(android))

10841
10840


### 刪除錯誤資料

開始分析前，我們必須確保資料內容正確，否則會影響分析結果。這表示我們必須：
- 偵測錯誤資料，修正或移除它
- 偵測重複資料，移除重複的部分

回想我們的公司是在做免費的App並上架供用戶下載，我們的客群是把英文的母語使用人士，因此我們必須移除：
- 移除非英文的App，例如<i>爱奇艺PPS -《欢乐颂2》电视剧热播</i>
- 移除付費App

這個步驟我們稱為<b>資料清理</b>，資料清理必須在分析之前完成，包括移除或校正錯誤資料、移除重複資料與調整資料以符合目的。

資料科學家通常會花80%的時間清理資料，再花20%的時間分析清理過的資料，在這項專案中，你會發現事實確實如此。


In [7]:
# 課程提示我們Instagram這個App重複地出現在資料中
# 我們先確認是否真的有重複

for app in android:
    name = app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


In [8]:
# 將重複的App名字append到空白的list
# 並計算重複個數

duplicate_apps = []
unique_apps = []

for app in android:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
    
print('重複資料數:', len(duplicate_apps))
print('\n')
print(duplicate_apps[:15])

重複資料數: 1181


['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


In [9]:
# 以下是我們預期清理後的資料數量

print('預期資料數', len(android)-1181)

預期資料數 9659


要移除重複的資料，我們必須要做的步驟是：

- 建立一個dictionary，而每一個key都是獨特的app名稱，取其value是最高瀏覽次數
- 使用此dictionary建立一個新的資料集，此時就不會有重複的app了


In [10]:
# 建立空白的dictionary
# 將瀏覽次數較高的資料留下
reviews_max = {}

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
        
    elif name not in reviews_max:
        reviews_max[name] = n_reviews

確認一下現在的資料筆數與預期的9659是否相同

In [11]:
print('預期資料數:', len(android) -len(duplicate_apps))
print('實際資料數:', len(reviews_max))

預期資料數: 9659
實際資料數: 9659


In [12]:
android_clean = []
already_added = []

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    
    if (reviews_max[name] == n_reviews) and (name not in already_added):
        android_clean.append(app)
        already_added.append(name) 

用函數explore_data()確認清理過的資料筆數是否正確

In [13]:
explore_data(android_clean, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


row數:  9659
column數:  13


前面的步驟我們移除了Google Play裡面重複的資料，此時我們要移除非英文的App。

此時我們將運用內建的ord()函數與ASCII(美國標準信息交換編碼)來判定app是否為英文，若為英文，ord()函數的值將介於0到127之間，若超過即為其他語系的文字。


In [14]:
# 課程提示這幾筆資料為非英文app

print(ios[813][1])
print(ios[6713][1])
print('\n')
print(android_clean[4412][0])
print(android_clean[7940][0])

爱奇艺PPS -《欢乐颂2》电视剧热播
東京時層地図 for iPad


中国語 AQリスニング
لعبة تقدر تربح DZ


In [15]:
# 建立一個函數辨別是否為英文
# 此時只是雛型

def is_english(string):
    
    for character in string:
        if ord(character) > 127:
            return False
    return True

In [16]:
# 用以下範例練習

print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))
print(ord('™'))
print(ord('😜'))

True
False
False
False
8482
128540


我們會發現如果App名稱中有表情符號或是特殊文字，我們建立的函數也會判定為「非英文」，這會使數據大量流失，所以我們要在函數中新增條件，要求出現三個以上的非英文字才判定為「非英文」

In [17]:
# 修正函數
def is_english(string):
    non_ascii = 0
    
    for character in string:
        if ord(character) > 127:
            non_ascii += 1
    
    if non_ascii > 3:
        return False
    else:
        return True

# 確認是否正確
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

True
True


In [18]:
is_english('爱奇艺PPS -《欢乐颂2》电视剧热播')

False

In [19]:
# 清理資料集
# is_english結果為True才能留下
android_english = []
ios_english = []

for app in android_clean:
    name = app[0]
    if is_english(name):
        android_english.append(app)
        
for app in ios:
    name = app[1]
    if is_english(name):
        ios_english.append(app)
        
explore_data(android_english, 0, 3, True)
print('\n')
explore_data(ios_english, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


row數:  9614
column數:  13


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5'

現在只剩9614個google apps和6183個iOS app

資料清理到目前為止，我們做了：
- 移除不正確的資料
- 移除重複的資料
- 移除非英文的資料

再次回想我們公司的產品是免費下載的app，主要獲利來源為app內的廣告收入，目前為止資料庫裡還包含免費與付費的資料，因此我們要再一次進行資料清理。

In [20]:
# 移除付費app的資料

android_final = []
ios_final = []

for app in android_english:
    price = app[7]
    if price == '0':
        android_final.append(app)
        
for app in ios_english:
    price = app[4]
    if price == '0.0':
        ios_final.append(app)
        
print('Android資料數:',len(android_final))
print('iOS資料數:',len(ios_final))

Android資料數: 8864
iOS資料數: 3222


目前為止，我們剩下8864筆Android資料與3222筆iOS資料

為了最小化風險與人力成本，我們的策略包含三個步驟：
1. 以最少成本在Android平台上架app
2. 如果用戶給予好的評價再進一步研發
3. 如果六個月後仍有獲利，我們再研發iOS版本與上架App Store

因為我們最終目標是在兩大平台都要上架，我們需要找到兩大平台都有上架的app，再作進一步的分析。

我們要建立兩個函數來分析各個app的種類的次數分配：
1. 產生次數分配並顯示百分比
2. 以遞減方式呈現百分比


In [21]:
def freq_table(dataset, index):
    """
    此函數為檢視資料庫內各app類別出現次數
    dataset為android或iOS
    index要看dataset所需檢視的欄位index為何
    像這次要檢視的genre在ios為[-5],在android為[-4]
    """
    table = {}
    total = 0
    
    for row in dataset:
        total += 1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
    
    table_percentages = {}
    for key in table:
        percentage = (table[key] / total) * 100
        table_percentages[key] = round(percentage,2)
    
    return table_percentages


def display_table(dataset, index):
    """
    運用前面寫的freq_table()函數算出個類別佔比
    再依序排列後印出結果
    """
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
        
    table_sorted = sorted(table_display, reverse = True)[:10] # 印出前十個即可
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [22]:
android_header

['App',
 'Category',
 'Rating',
 'Reviews',
 'Size',
 'Installs',
 'Type',
 'Price',
 'Content Rating',
 'Genres',
 'Last Updated',
 'Current Ver',
 'Android Ver']

### 檢視iOS平台免費app的Prime genre組成

In [23]:
# 要檢視的prime genre的index為-5
display_table(ios_final, -5)

Games : 58.16
Entertainment : 7.88
Photo & Video : 4.97
Education : 3.66
Social Networking : 3.29
Shopping : 2.61
Utilities : 2.51
Sports : 2.14
Music : 2.05
Health & Fitness : 2.02


我們可以發現在ios平台遊戲類app佔大多數約58%，其次為娛樂類約8%，緊接照片&影片類別約5%，教育類與社交類別分別佔約3.7%與3.3%


### 檢視Google Play平台免費app的category組成

In [34]:
display_table(android_final, 1)

FAMILY : 18.91
GAME : 9.72
TOOLS : 8.46
BUSINESS : 4.59
LIFESTYLE : 3.9
PRODUCTIVITY : 3.89
FINANCE : 3.7
MEDICAL : 3.53
SPORTS : 3.4
PERSONALIZATION : 3.32


我們可以發現google play的結構與ios相當不同，家庭類app為最大宗約18.9%，遊戲類位居第二約9.7%，第三為工具類8.5%，緊接在後是商業類4.5%，生活類3.9%。

可以發現google play的軟體偏向實用層面，不過深入研究會發現，家庭類別其實也包含許多給小孩遊玩的遊戲，因此遊戲類還是佔多數。

[Google play家庭類app](https://play.google.com/store/apps/category/FAMILY?hl=en)

### 檢視Google Play平台的genre占比情形

In [25]:
display_table(android_final, -4) # genre

Tools : 8.45
Entertainment : 6.07
Education : 5.35
Business : 4.59
Productivity : 3.89
Lifestyle : 3.89
Finance : 3.7
Medical : 3.53
Sports : 3.46
Personalization : 3.32


在Google Play平台的category和genre定義比較不明確，但我們可以確定的是category欄位比較大方向，而genre欄位比較細分，而目前我們只需要觀察較大面向的趨勢，所以我們使用category欄位即可。

目前為止，我們發現App Store被遊戲類的app統治，大約58%都是遊戲類；而Google Play則在實用與樂趣之間取得平衡。

因此，為了要在兩大平台都上架產品，我們會建議研發部門針對遊戲類別下手開發，先在Google Play平台上架後蒐集用戶反饋，按照反饋修正，若連續六個月都能獲利，我們研發在iOS平台上架。


## 假設我們也考慮切入收費app領域並從中獲利，我們要切入哪一類的app？

In [30]:
# 移除免費app

android_nonfree = []
ios_nonfree = []

for app in android_english:
    price = app[7]
    if price != '0':
        android_nonfree.append(app)
        
for app in ios_english:
    price = app[4]
    if price != '0.0':
        ios_nonfree.append(app)

In [46]:
print('Android付費app資料數:',len(android_nonfree), '付費app比率:',round(len(android_nonfree)/len(android_english)*100,2),'%')
print('iOS付費app資料數:',len(ios_nonfree),'付費app比率:',round(len(ios_nonfree)/len(ios_english)*100,2),'%')

Android付費app資料數: 750 付費app比率: 7.8 %
iOS付費app資料數: 2961 付費app比率: 47.89 %


在清理過的抽樣數據中，Android僅有約8%的付費app，而iOS平台有將近48%的付費app，如果公司想要透過付費app獲利，建議先上架iOS平台，畢竟用戶在iOS平台比較常見付費app，先透過累積iOS平台的用戶與反饋修正app，再根據進階數據研擬是否要上架Android平台。

### 檢視iOS平台付費app的Prime genre組成

In [33]:
display_table(ios_nonfree, -5)

Games : 51.27
Education : 9.86
Entertainment : 6.59
Photo & Video : 6.11
Utilities : 4.46
Productivity : 3.78
Health & Fitness : 3.38
Music : 2.4
Lifestyle : 1.62
Weather : 1.38


即使目前條件改為付費app，iOS平台仍有高達51%的遊戲類app，不過教育類提升至接近10%，若公司原本就有研發教育類的app且著手進入付費app市場，除了最大宗的遊戲類之外，還可以考慮研發教育類app。

### 檢視Google Play平台付費app的category組成

In [35]:
display_table(android_nonfree, 1)

FAMILY : 24.27
MEDICAL : 10.93
GAME : 10.93
PERSONALIZATION : 10.8
TOOLS : 10.4
PRODUCTIVITY : 3.73
BOOKS_AND_REFERENCE : 3.73
COMMUNICATION : 3.6
SPORTS : 3.2
PHOTOGRAPHY : 2.53


我們在Android平台可以發現比較大的不同，將條件改為付費app後，占比最大的種類改為家庭(24%)、醫療(11%)、遊戲、個人化與工具(各10%)，Android平台付費app比較常出現在功能性領域，若公司決定先上架iOS平台再上架Android平台，家庭與遊戲類都是在兩大平台占比較前面的類別，都可以考慮由此著手。